src/gui/guiWrapper.h

```diff
@@ -41,14 +43,44 @@ struct WindowInfo
 	std::atomic_bool has_screenshot_request;
 	std::atomic_bool is_fullscreen;
 
-	std::atomic_bool keydown[256]{};
-
+	void set_keystate(uint32 keycode, bool state)
+	{
+		const std::lock_guard<std::mutex> lock(keycode_mutex);
+		m_keydown[keycode] = state;
+	};
+	bool get_keystate(uint32 keycode)
+	{
+		const std::lock_guard<std::mutex> lock(keycode_mutex);
+		auto result = m_keydown.find(keycode);
+		if (result == m_keydown.end())
+			return false;
+		return result->second;
+	}
+	std::unordered_map<uint32, bool> get_keystates() const
```

Comment:
Creating a temporary copy of the entire container is pretty inefficient and in this form it's also not thread-safe.
A better design would be to pass the target map for the copy as a reference, like this:
`void get_keystates(std::unordered_map<uint32, bool>& buttonsOut)`
Then inside the function you can handle the copying/assignment while also locking `keycode_mutex`.